tests/test_project/apps.py

```diff
@@ -1,18 +1,33 @@
 from django.apps import AppConfig
 from django.conf import settings
 
+from openwisp_utils import settings as app_settings
+
 
 class TestAppConfig(AppConfig):
     name = 'test_project'
     label = 'test_project'
 
+    _default_drf = {
+        'DEFAULT_THROTTLE_CLASSES': ['rest_framework.throttling.ScopedRateThrottle'],
+        'DEFAULT_THROTTLE_RATES': {
+            'anon': app_settings.API_RATE_ANON,
+        },
+    }
+
     def ready(self, *args, **kwargs):
         super(TestAppConfig, self).ready(*args, **kwargs)
         self.add_default_menu_items()
+        self.configure_drf_defaults()
 
     def add_default_menu_items(self):
         menu_setting = 'OPENWISP_DEFAULT_ADMIN_MENU_ITEMS'
         items = [
             {'model': 'test_project.Shelf'},
         ]
         setattr(settings, menu_setting, items)
+
+    def configure_drf_defaults(self):
+        config = getattr(settings, 'REST_FRAMEWORK', {})
+        config.update(self._default_drf)
+        setattr(settings, 'REST_FRAMEWORK', config)
```

Comment:
It looks to me that this way of doing it does not allow users to override our defaults.

I tried putting this in test/settings.py:

REST_FRAMEWORK = {'DEFAULT_THROTTLE_RATES': {'anon': '100/hour'}}

But in the django shell I see `40/hour:

$ ./manage.py shell
In [1]: from django.conf import settings                                                                                                                                                                           

In [2]: settings.REST_FRAMEWORK                                                                                                                                                                                    
Out[2]: 
{'DEFAULT_THROTTLE_CLASSES': ['rest_framework.throttling.ScopedRateThrottle'],
 'DEFAULT_THROTTLE_RATES': {'anon': '40/hour'}}

I think it would be better to allow users to override our defaults, we could do something like the following:

- does the project settings file define `REST_FRAMEWORK['DEFAULT_THROTTLE_CLASSES']`? If yes, don't override it
- before setting `self._default_drf['DEFAULT_THROTTLE_RATES']` check if each key is already set in `REST_FRAMEWORK['DEFAULT_THROTTLE_RATES']`, if it is, skip

With this approach we would be able to:

- supply our own default values
- allow users to override them
- avoid the need of defining settings like `OPENWISP_API_RATE_ANON`, because they can just do:

```python
REST_FRAMEWORK = {'DEFAULT_THROTTLE_RATES': { 'anon': '100/hour'}}
```

This way of doing this is also familiar to many django developers and is also [documented in the DRF docs](https://www.django-rest-framework.org/api-guide/throttling/#setting-the-throttling-policy), which we could link from our own docs, so we would be following a standard that is already well known in the community.

What do you think?

Reply:
Can't agree more!